plugins/com.google.cloud.tools.eclipse.appengine.compat/src/com/google/cloud/tools/eclipse/appengine/compat/GpeMigrator.java

```diff
@@ -108,29 +120,35 @@ private static boolean isGpeClasspath(IClasspathEntry entry) {
   }
 
   @VisibleForTesting
-  static void removeGpeNature(IProject project) {
+  static boolean removeGpeNature(IProject project) {
     try {
+      boolean hadNature = NatureUtils.hasNature(project, GPE_GAE_NATURE_ID);
       NatureUtils.removeNature(project, GPE_GAE_NATURE_ID);
+      return hadNature;
     } catch (CoreException ex) {
       logger.log(Level.WARNING, "Failed to remove GPE nature.", ex);
+      return true;
     }
   }
 
   @VisibleForTesting
-  static void removeGpeRuntimeAndFacets(IFacetedProject facetedProject) {
+  static boolean removeGpeRuntimeAndFacets(IFacetedProject facetedProject) {
     // To remove the facets, we will directly modify the WTP facet metadata file (using XSLT):
     // .settings/org.eclipse.wst.common.project.facet.core.xml
     IFile metadataFile = facetedProject.getProject().getFile(FACETS_METADATA_FILE);
     if (!metadataFile.exists()) {
-      return;
+      return false;
     }
-
+    // must make a copy
+    Set<IProjectFacetVersion> originalFacets = new HashSet<>(facetedProject.getProjectFacets());
```

Comment:
They're not big, we have to compute the results anyways, and it helped when debugging.